### PR TITLE
[fix](compute_node) fix backend_num_for_federation bug for federation…

### LINKED ISSUE #13144: [Enhancement] Support Compute Only Node in doris
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

As [discussion](https://docs.google.com/document/d/1VafP0gu5o8WCYaII1mDjElseaJEOnKF0Ai4bxDIfVQE/edit) in the Dev meeting, I'm propose the Compute Only Node in doris.
A Compute Node(CN) will be different from BE as these aspects:
1. No any storeage in CN node. These will be no any tablet schedule to any CN node.
2. Only support the federation query such as hms table and es table.
3. Can be asigned the plan node except the table scan node when the workload is heavy in the be node.
4. Can be added or delete immediately.

### Solution

In my idea, I will impl this feature as these step:
1. Introduce the node role to doris, and the table creation and tablet scheduler will control the storage only assign to the BE nodes.
2.  Support the olap query plan running on the CN node when  workload is heavy
3. Impl the graceful shutdown when the doris pipeline engine is ready

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by nextdreamblue**
… query

Signed-off-by: nextdreamblue <zxw520blue1@163.com>

# Proposed changes

Issue Number: close #13144

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [x] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.38 seconds
 load time: 471 seconds
 storage size: 17170856498 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230207065033_clickbench_pr_91195.html

File: fe/fe-common/src/main/java/org/apache/doris/common/Config.java [hunk 1]
```diff
@@ -1819,6 +1819,8 @@ public class Config extends ConfigBase {
      * The candidate of the backend node for federation query such as hive table and es table query.
      * If the backend of computation role is less than this value, it will acquire some mix backend.
      * If the computation backend is enough, federation query will only assign to computation backend.
+     * If value = 0, will get none backends for federation query.
```
Comment (SaintBacchus) [inline fe/fe-common/src/main/java/org/apache/doris/common/Config.java:1822]: Has `backend_num_for_federation` any meaning?

Should we treat 0 be the same as -1?

File: docs/en/docs/advanced/compute_node.md [hunk 2]
```diff
@@ -96,15 +96,17 @@ HeartbeatFailureCounter: 0
 ```
 
 ### Usage
-When using the [MultiCatalog](../lakehouse/multi-catalog/multi-catalog) , the query will be preferentially scheduled to the compute node.
+When using the [MultiCatalog](https://doris.apache.org/docs/dev/lakehouse/multi-catalog/) , the query will be preferentially scheduled to the compute node.
```
Comment (SaintBacchus) [inline docs/en/docs/advanced/compute_node.md:99]: It seems had been fix by #16284 but the doc is not released

File: fe/fe-common/src/main/java/org/apache/doris/common/Config.java (hunk 1 above)
Comment (nextdreamblue) [reply fe/fe-common/src/main/java/org/apache/doris/common/Config.java:1822]: 也算是多一个不同的选项吧，0返回空，也和一般的理解相符。如果没必要提供这类含义，我可以改下代码

File: docs/en/docs/advanced/compute_node.md (hunk 2 above)
Comment (nextdreamblue) [reply docs/en/docs/advanced/compute_node.md:99]: ok, i revert this code

**Comment by SaintBacchus** [comment]
LGTM

**Comment by nextdreamblue** [comment]
https://github.com/apache/doris/pull/16737
这个pr把这块的逻辑完全改了，我这个pr已经没有意义了，close了

**Pull request closed without merging**